### PR TITLE
[SHTns] bump v3.5.2, allow omp threads

### PR DESCRIPTION
**Pull request by fgerick**
also hack around fftw_omp requirement of shtns to allow SHTns_jll to use FFTW_jll threads.

### CODE REVIEW

File: S/SHTns/build_tarballs.jl [hunk 1]
```diff
@@ -3,44 +3,48 @@
...
-./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target}
-if [[ "${target}" == *-apple-* ]]; then
-    # For some reasons, `configure` insists on setting CC2=gcc, also on macOS
-    sed -i -e 's/gcc/cc/' -e 's/ -fno-tree-loop-distribute-patterns//' Makefile
-fi
+export CC="gcc"
+export CFLAGS="-fPIC -O3" #only -fPIC produces slow code on linux x86 and MacOS x86 (maybe others)
+# I remove openmp fftw checks, because FFTW_jll doesn't provide -lfftw3_omp
+# propaly not the safest way!
+sed -i -e '4908,4957d' configure
```
Comment (giordano) [inline S/SHTns/build_tarballs.jl:20]: Playing with line numbers is a terrible way to deal with this. It's hard to tell what you're doing and if anything changes in a following release the risk of screwing things up badly is non-zero. Applying a proper patch is a safer and more explicit way to...patch a file.

File: S/SHTns/build_tarballs.jl [hunk 2]
```diff
@@ -3,44 +3,48 @@
...
+# I remove openmp fftw checks, because FFTW_jll doesn't provide -lfftw3_omp
+# propaly not the safest way!
+sed -i -e '4908,4957d' configure 
+./configure --prefix=${prefix} --host=${target} --enable-openmp
 make -j${nproc}
 make install
 mkdir -p ${libdir}
-cc -shared -o "${libdir}/libshtns.${dlext}" *.o -lfftw3
-rm "${prefix}/lib/libshtns.a"
+gcc -fopenmp -shared -o "${libdir}/libshtns.${dlext}" *.o -lfftw3
```
Comment (giordano) [inline S/SHTns/build_tarballs.jl:25]: Why gcc explicitly?

File: S/SHTns/build_tarballs.jl [hunk 3]
```diff
@@ -3,44 +3,48 @@
...
 # Bash recipe for building across all platforms
 script = raw"""
 cd $WORKSPACE/srcdir/shtns/
-export CFLAGS="-fPIC"
-./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target}
-if [[ "${target}" == *-apple-* ]]; then
-    # For some reasons, `configure` insists on setting CC2=gcc, also on macOS
-    sed -i -e 's/gcc/cc/' -e 's/ -fno-tree-loop-distribute-patterns//' Makefile
-fi
+export CC="gcc"
```
[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl:16]: without explicit gcc I get e.g.
```
ld: library not found for -lomp
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
Previous command exited with 1
ERROR: LoadError: Build for SHTns on aarch64-apple-darwin did not complete successfully
```

File: S/SHTns/build_tarballs.jl (hunk 2 above)
Comment (fgerick) [reply S/SHTns/build_tarballs.jl:25]: Is there a better way of making openmp threads work on all architectures?

File: S/SHTns/build_tarballs.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl:16]: https://github.com/JuliaPackaging/Yggdrasil/blob/83900358eee9ba66384031faecdd8edde2d27029/B/blis/build_tarballs.jl#L135-L138

File: S/SHTns/build_tarballs.jl [hunk 4]
```diff
@@ -3,44 +3,48 @@
...
+    Dependency(PackageSpec(name="FFTW_jll", uuid="f5851436-0d7a-5f13-b9de-f02708fd171a")),
+    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae"))
 ]
 
 # Build the tarballs, and possibly a `build.jl` as well.
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6")
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; 
+                    julia_compat="1.6", 
+                    # preferred_gcc_version=v"12", 
+                    lock_microarchitecture=false,
```
[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl:49]: `-march=native` is ***never*** a good way to compile code to be used on other machines.

[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl:49]: Would it be better to include `CFLAGS` depending on the cpu, similar to what is done here?:
https://github.com/JuliaPackaging/Yggdrasil/blob/9b47a018c828d07c801223fec00df12da4c9589e/F/FFTW/build_tarballs.jl#L29

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl:49]: FFTW uses CPUID though, which I don't think this package does.

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl:49]: Yeah, that's a bit better

File: S/SHTns/build_tarballs.jl [hunk 5]
```diff
@@ -3,44 +3,52 @@
...
 
 # Bash recipe for building across all platforms
 script = raw"""
 cd $WORKSPACE/srcdir/shtns/
-export CFLAGS="-fPIC"
-./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target}
-if [[ "${target}" == *-apple-* ]]; then
-    # For some reasons, `configure` insists on setting CC2=gcc, also on macOS
-    sed -i -e 's/gcc/cc/' -e 's/ -fno-tree-loop-distribute-patterns//' Makefile
-fi
```
Comment (giordano) [inline S/SHTns/build_tarballs.jl:21]: Was that fixed?

Comment (giordano) [reply S/SHTns/build_tarballs.jl:21]: Oh, that probably can be fixed with `--enable-kernel-compiler=cc`

File: S/SHTns/build_tarballs.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl]: I have added this into build file. For avx512 there were some missing symbols in gcc-6, when I select the preferred gcc to be 12 it builds these fine. I did a quick benchmark on an intel xeon processor and the binary build using the gcc-12 with avx512 seems to be as fast/faster compared to the native build on the machine 🚀

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl]: Please don't use gcc 12. If there's a problem with gcc 6, go up from there, but don't start from the top. GCC has many compatibility problems.

[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl]: is there a reason to avoid `preferred_gcc_version=v"12"` ?

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl]: Don't use the number 12? Again, if you had problems with gcc 6, try 7, don't jump to 12 directly.

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl]: What's the problem with gcc 6 BTW? We might just force gcc 7 for AVX512 if that's a serious issue

File: S/SHTns/build_tarballs.jl (hunk 5 above)
Comment (fgerick) [reply S/SHTns/build_tarballs.jl:21]: yes, that build was before your comment, without the flag

File: S/SHTns/build_tarballs.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl]: ok I'll try to find the smallest possible version

[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl]: This is what i get without any gcc preference.
```
cc -fPIC -O3 -I/workspace/destdir/include -L/workspace/destdir/lib -mtune=skylake -std=gnu99 -fopenmp -D_GNU_SOURCE -DSHTNS_GIT=\"\" -DSHTNS_VER=\"3.5.2\"  -c sht_fly.c -o sht_fly.o
In file included from sht_private.h:239:0,
                 from sht_init.c:31:
shtns_simd.h: In function ‘reduce_add’:
shtns_simd.h:258:11: warning: implicit declaration of function ‘_mm512_reduce_add_pd’ [-Wimplicit-function-declaration]
    return _mm512_reduce_add_pd(a);
           ^~~~~~~~~~~~~~~~~~~~
In file included from sht_private.h:239:0,
                 from sht_odd_nlat.c:1:
shtns_simd.h: In function ‘reduce_add’:
shtns_simd.h:258:11: warning: implicit declaration of function ‘_mm512_reduce_add_pd’ [-Wimplicit-function-declaration]
    return _mm512_reduce_add_pd(a);
           ^~~~~~~~~~~~~~~~~~~~
In file included from sht_private.h:239:0,
                 from sht_kernels_a.c:18:
shtns_simd.h: In function ‘reduce_add’:
shtns_simd.h:258:11: warning: implicit declaration of function ‘_mm512_reduce_add_pd’ [-Wimplicit-function-declaration]
    return _mm512_reduce_add_pd(a);
           ^~~~~~~~~~~~~~~~~~~~
In file included from sht_private.h:239:0,
                 from sht_omp.c:18:
shtns_simd.h: In function ‘reduce_add’:
shtns_simd.h:258:11: warning: implicit declaration of function ‘_mm512_reduce_add_pd’ [-Wimplicit-function-declaration]
    return _mm512_reduce_add_pd(a);
           ^~~~~~~~~~~~~~~~~~~~
In file included from sht_private.h:239:0,
                 from sht_kernels_s.c:18:
shtns_simd.h: In function ‘reduce_add’:
shtns_simd.h:258:11: warning: implicit declaration of function ‘_mm512_reduce_add_pd’ [-Wimplicit-function-declaration]
    return _mm512_reduce_add_pd(a);
           ^~~~~~~~~~~~~~~~~~~~
In file included from sht_private.h:239:0,
                 from sht_fly.c:18:
shtns_simd.h: In function ‘reduce_add’:
shtns_simd.h:258:11: warning: implicit declaration of function ‘_mm512_reduce_add_pd’ [-Wimplicit-function-declaration]
    return _mm512_reduce_add_pd(a);
           ^~~~~~~~~~~~~~~~~~~~
In file included from sht_fly.c:54:0:
SHT/SH_to_spat_fly.c: In function ‘SH_m_to_spat_fly4_l’:
SHT/SH_to_spat_fly.c:315:3: internal compiler error: in simplify_binary_operation_1, at simplify-rtx.c:3516
   }
   ^
0x9c865e simplify_binary_operation_1
	/workspace/srcdir/gcc-6.1.0/gcc/simplify-rtx.c:3516
0x9c8715 simplify_binary_operation(rtx_code, machine_mode, rtx_def*, rtx_def*)
	/workspace/srcdir/gcc-6.1.0/gcc/simplify-rtx.c:2063
0xfa6455 fold_rtx
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:3728
0xfa628a fold_rtx
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:3228
0xfa7619 cse_insn
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:4639
0xfac013 cse_extended_basic_block
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:6535
0xfac013 cse_main
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:6709
0xfac9c9 rest_of_handle_cse
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:7529
0xfac9c9 execute
	/workspace/srcdir/gcc-6.1.0/gcc/cse.c:7572
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://gcc.gnu.org/bugs.html> for instructions.
make: *** [Makefile:71: sht_fly.o] Error 1
make: *** Waiting for unfinished jobs....
Previous command exited with 2
ERROR: LoadError: Build for SHTns on x86_64-linux-gnu-march+avx512 did not complete successfully
```

what is the elegant way to force gcc 7 for avx512 only? separating the `platforms` ? would the augmentation work the same way?

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl]: > This is what i get without any gcc preference.

Ah, that's https://github.com/JuliaPackaging/Yggdrasil/pull/4485#issuecomment-1048226993.


> what is the elegant way to force gcc 7 for avx512 only? separating the platforms ? would the augmentation work the same way?

There's a way, but it isn't very elegant.  Using GCC 7 everywhere is fine.

File: S/SHTns/build_tarballs.jl [hunk 6]
```diff
@@ -2,35 +2,52 @@
...
+
 install_license LICENSE
 """
 
 # These are the platforms we will build for by default, unless further
 # platforms are passed in on the command line
-platforms = supported_platforms()
 
+# Expand for microarchitectures on x86_64 (library doesn't have CPU dispatching)
+platforms = expand_microarchitectures(supported_platforms(), ["x86_64", "avx", "avx2", "avx512"])
```
[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl:38]: x86-64-w64-mingw32-march-plus-avx51 does compile with gcc-9. Should I separate and build x86 windows with gcc-9?

[the remark this answers is not in the document]
Comment (giordano) [reply S/SHTns/build_tarballs.jl:38]: GCC 9 is still ok, this is also _not_ a C++ code, which would be much more problematic in terms of compatibility (it'd require a matching libstdc++ runtime).

File: S/SHTns/build_tarballs.jl [hunk 7]
```diff
@@ -39,8 +56,16 @@ products = [
...
+    Dependency(PackageSpec(name="LLVMOpenMP_jll", uuid="1d63c593-3942-5779-bab2-d838dc0a180e"); platforms=filter(Sys.isbsd, platforms)),
 ]
 
 # Build the tarballs, and possibly a `build.jl` as well.
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6")
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; 
+                julia_compat="1.6", 
+                preferred_gcc_version=v"7",
+                augment_platform_block)
+
```
Comment (giordano) [inline S/SHTns/build_tarballs.jl:71]:
```suggestion
                preferred_gcc_version=v"8",
                augment_platform_block)
```

**Comment by giordano** [comment]
Side note, as a general rule working with Git and GitHub, opening a PR from a branch with the same name as the target branch is [anti-pattern](https://blog.jasonmeridth.com/posts/do-not-issue-pull-requests-from-your-master-branch/).  You should always create a ***new branch***.

**Comment by fgerick** [comment]
Looks good to me! Thank you for the support and I learned a lot with this PR. Open source contribution is not on my daily menu and there's a bit to be learned still. Thanks again!

File: S/SHTns/build_tarballs.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (fgerick) [reply S/SHTns/build_tarballs.jl:38]: You're right I was already at v10 when it didn't fail anymore!

**Pull request merged**